### PR TITLE
Refactor trainers classes to use BaseTrainer with shared functionality

### PR DESCRIPTION
**Pull request by albertvillanova**
Refactor trainers classes to use `BaseTrainer` with shared functionality. Concretely:
- Factorize trainers `create_model_card`

This PR refactors the trainer classes to improve code reuse and maintainability, resulting in cleaner and more consistent code across the trainers. The main change is the introduction of a new `BaseTrainer` class, which centralizes shared functionality. All existing trainer classes now inherit from `BaseTrainer` to delegate shared functionalities:
- Logic for model card creation

### Changes
- Trainer class refactoring:
  - Add a new `BaseTrainer` class in `trl/trainer/base_trainer.py` that inherits from `Trainer` and provides a shared `create_model_card` method for generating and saving model cards.
  - Update all trainer classes to inherit from `BaseTrainer` instead of `Trainer`, and removed duplicated model card creation logic from each trainer.
- Imports and code cleanup:
  - Remove redundant imports (`Trainer`, `generate_model_card`, `get_comet_experiment_url`, and others) from trainer files, since these are now handled by `BaseTrainer`.
- Model card creation delegation:
  - Remove the `create_model_card` method in each trainer to delegate to `BaseTrainer.create_model_card`, simplifying the code and ensuring consistent tag normalization and metadata handling.
  - Add `_name` and `_paper` class attributes to all trainers with custom data for `create_model_card`
- Add missing `paper_id` to `PRMTrainer`.

This refactor makes the trainer codebase easier to maintain and extend, as shared logic is now centralized and duplicated code has been removed.

Example of code duplication and maintainability burden:
- #4092

### CODE REVIEW

**Comment by qgallouedec** [comment]
lgtm! For the future, I think this class will remain fairly light, because if a feature is common to all trainers, then it would probably make sense to integrate it directly into `transformers.Trainer`. `create_model_card` is an exception in this regard.

**Comment by albertvillanova** [comment]
Thanks for your fast review! I take note of the insight: this base class can give us hints about what would be better to integrate in transformers.

By the way, I think the errors are unrelated?

I am just working on an additional factorization and the PR will be finished.

**Comment by albertvillanova** [comment]
@qgallouedec, could you approve this?
The failing CI tests are unrelated.

File: trl/trainer/base_trainer.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+
+class BaseTrainer(Trainer):
+    _tag_names = []
+    _name = "Base"
+    _paper = {
+        "title": None,
+        "id": None,
+        "citation": None,
+    }
```
Comment (qgallouedec) [inline trl/trainer/base_trainer.py]:
```suggestion
    _paper = {}
```

File: trl/trainer/sft_trainer.py [hunk 2]
```diff
@@ -590,6 +585,8 @@ class SFTTrainer(Trainer):
     """
 
     _tag_names = ["trl", "sft"]
+    _name = "SFT"
+    _paper = {}
```
Comment (qgallouedec) [inline trl/trainer/sft_trainer.py]:
```suggestion
```

File: trl/trainer/reward_trainer.py [hunk 3]
```diff
@@ -123,6 +116,8 @@ class RewardTrainer(Trainer):
     """
 
     _tag_names = ["trl", "reward-trainer"]
+    _name = "Reward"
+    _paper = {}
```
Comment (qgallouedec) [inline trl/trainer/reward_trainer.py]:
```suggestion
```

**Review by qgallouedec — COMMENTED** [review]
what do you think of this instead?
